pallets/dapp-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,2028 @@
+            Ok(())
+        }
+
+        /// Used to register a new contract for dApp staking.
+        ///
+        /// If successful, smart contract will be assigned a simple, unique numerical identifier.
+        /// Owner is set to be initial beneficiary & manager of the dApp.
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::register())]
+        pub fn register(
+            origin: OriginFor<T>,
+            owner: T::AccountId,
+            smart_contract: T::SmartContract,
+        ) -> DispatchResult {
+            Self::ensure_pallet_enabled()?;
+            T::ManagerOrigin::ensure_origin(origin)?;
+
+            ensure!(
+                !IntegratedDApps::<T>::contains_key(&smart_contract),
+                Error::<T>::ContractAlreadyExists,
+            );
+
+            ensure!(
+                IntegratedDApps::<T>::count() < T::MaxNumberOfContracts::get().into(),
+                Error::<T>::ExceededMaxNumberOfContracts
+            );
+
+            let dapp_id = NextDAppId::<T>::get();
+            // MAX value must never be assigned as a dApp Id since it serves as a sentinel value.
+            ensure!(dapp_id < DAppId::MAX, Error::<T>::NewDAppIdUnavailable);
+
+            IntegratedDApps::<T>::insert(
+                &smart_contract,
+                DAppInfo {
+                    owner: owner.clone(),
+                    id: dapp_id,
+                    state: DAppState::Registered,
+                    reward_destination: None,
+                },
+            );
+
+            NextDAppId::<T>::put(dapp_id.saturating_add(1));
+
+            Self::deposit_event(Event::<T>::DAppRegistered {
+                owner,
+                smart_contract,
+                dapp_id,
+            });
+
+            Ok(())
+        }
+
+        /// Used to modify the reward beneficiary account for a dApp.
+        ///
+        /// Caller has to be dApp owner.
+        /// If set to `None`, rewards will be deposited to the dApp owner.
+        /// After this call, all existing & future rewards will be paid out to the beneficiary.
+        #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::set_dapp_reward_beneficiary())]
+        pub fn set_dapp_reward_beneficiary(
+            origin: OriginFor<T>,
+            smart_contract: T::SmartContract,
+            beneficiary: Option<T::AccountId>,
+        ) -> DispatchResult {
+            Self::ensure_pallet_enabled()?;
+            let dev_account = ensure_signed(origin)?;
+
+            IntegratedDApps::<T>::try_mutate(
+                &smart_contract,
+                |maybe_dapp_info| -> DispatchResult {
+                    let dapp_info = maybe_dapp_info
+                        .as_mut()
+                        .ok_or(Error::<T>::ContractNotFound)?;
+
+                    ensure!(dapp_info.owner == dev_account, Error::<T>::OriginNotOwner);
+
+                    dapp_info.reward_destination = beneficiary.clone();
```

Comment:
Maybe rename `reward_destination` to `reward_beneficiary` for consistency? The `DAppRewardDestinationUpdated` error uses `Destination` as well.